pkg/mcs/resource_manager/client/limiter.go

```diff
@@ -0,0 +1,401 @@
+// A Reservation may be canceled, which may enable the Limiter to permit additional events.
+type Reservation struct {
+	ok        bool
+	lim       *Limiter
+	tokens    float64
+	timeToAct time.Time
+	// This is the Limit at reservation time, it can change later.
+	limit Limit
+}
+
+// OK returns whether the limiter can provide the requested number of tokens
+// within the maximum wait time.  If OK is false, Delay returns InfDuration, and
+// Cancel does nothing.
+func (r *Reservation) OK() bool {
+	return r.ok
+}
+
+// Delay is shorthand for DelayFrom(time.Now()).
+func (r *Reservation) Delay() time.Duration {
+	return r.DelayFrom(time.Now())
+}
+
+// InfDuration is the duration returned by Delay when a Reservation is not OK.
+const InfDuration = time.Duration(1<<63 - 1)
+
+// DelayFrom returns the duration for which the reservation holder must wait
+// before taking the reserved action.  Zero duration means act immediately.
+// InfDuration means the limiter cannot grant the tokens requested in this
+// Reservation within the maximum wait time.
+func (r *Reservation) DelayFrom(now time.Time) time.Duration {
+	if !r.ok {
+		return InfDuration
+	}
+	delay := r.timeToAct.Sub(now)
+	if delay < 0 {
+		return 0
+	}
+	return delay
+}
+
+// CancelAt indicates that the reservation holder will not perform the reserved action
+// and reverses tokens which be refilled into limiter.
+func (r *Reservation) CancelAt(now time.Time) {
+	if !r.ok {
+		return
+	}
+
+	r.lim.mu.Lock()
+	defer r.lim.mu.Unlock()
+
+	if r.lim.limit == Inf || r.tokens == 0 || r.timeToAct.Before(now) {
+		return
+	}
+	// advance time to now
+	now, _, tokens := r.lim.advance(now)
+	// calculate new number of tokens
+	tokens += float64(r.tokens)
+
+	// update state
+	r.lim.last = now
+	r.lim.tokens = tokens
+}
+
+// Reserve returns a Reservation that indicates how long the caller must wait before n events happen.
+// The Limiter takes this Reservation into account when allowing future events.
+// The returned Reservation’s OK() method returns false if waitting duration exceeds deadline.
+// Usage example:
+//
+//	r := lim.Reserve(time.Now(), 1)
+//	if !r.OK() {
+//	  // Not allowed to act! Did you remember to set lim.burst to be > 0 ?
+//	  return
+//	}
+//	time.Sleep(r.Delay())
+//	Act()
+//
+// Use this method if you wish to wait and slow down in accordance with the rate limit without dropping events.
+func (lim *Limiter) Reserve(ctx context.Context, now time.Time, n float64) *Reservation {
+	// Check if ctx is already cancelled
+	select {
+	case <-ctx.Done():
+		return &Reservation{
+			ok: false,
+		}
+	default:
+	}
+	// Determine wait limit
+	waitLimit := InfDuration
+	if deadline, ok := ctx.Deadline(); ok {
+		waitLimit = deadline.Sub(now)
+	}
+	r := lim.reserveN(now, n, waitLimit)
+	return &r
+}
+
+// SetupNotificationAt enables the notification at the given threshold.
+func (lim *Limiter) SetupNotificationAt(now time.Time, threshold float64) {
```

Comment:
Using the word `At` is a little bit confusing for the caller. They may be confused about which parameter this `At` refers to.

```suggestion
func (lim *Limiter) SetupNotificationThreshold(now time.Time, threshold float64) {
```